server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java

```diff
@@ -70,18 +70,40 @@ private static String valueToString(Object value) {
     @Override
     public final Query termQuery(Object value, QueryShardContext context) {
         String pattern = valueToString(value);
-        if (matches(pattern, context)) {
+        if (matches(pattern, false, context)) {
             return Queries.newMatchAllQuery();
         } else {
             return new MatchNoDocsQuery();
         }
     }
 
+    @Override
+    public final Query termQueryCI(Object value, QueryShardContext context) {
+        String pattern = valueToString(value);
+        if (matches(pattern, true, context)) {
+            return Queries.newMatchAllQuery();
+        } else {
+            return new MatchNoDocsQuery();
+        }
+    }
+    
     @Override
     public final Query termsQuery(List<?> values, QueryShardContext context) {
         for (Object value : values) {
             String pattern = valueToString(value);
-            if (matches(pattern, context)) {
+            if (matches(pattern, false, context)) {
+                // `terms` queries are a disjunction, so one matching term is enough
+                return Queries.newMatchAllQuery();
+            }
+        }
+        return new MatchNoDocsQuery();
+    }    
+
+    @Override
+    public final Query termsQueryCI(List<?> values, QueryShardContext context) {
```

Comment:
what do we gain shortening CaseInsensitive to CI here?

Reply:
Was the suggestion from @jimczi  - I'm not averse to changing. Jim?